### PR TITLE
fixed Q-Coherence chip

### PR DESCRIPTION
**Pull request by Kubik161**
Fix for #4611 and some related tests.

Review my changes please. I'm not sure it's correct fix, but it seems to work. Thanks.

### CODE REVIEW

File: src/clj/game/cards/hardware.clj [hunk 1]
```diff
@@ -1356,7 +1356,9 @@
    {:in-play [:memory 1]
     :events (let [e {:async true
                      :interactive (req true)
-                     :req (req (some program? targets))
+                     :req (req (some #(and (installed? %)
+                                           (program? %))
+                                   targets))
```
Comment (lostgeek) [inline src/clj/game/cards/hardware.clj]:
```suggestion
                                     targets))
```

File: test/clj/game_test/cards/hardware.clj [hunk 2]
```diff
@@ -2217,6 +2217,55 @@
...
+      (take-credits state :corp)
+      (play-from-hand state :runner "Q-Coherence Chip")
+      (play-from-hand state :runner "Self-modifying Code")
+      (let [smc1 (get-program state 0),
+            qchip (get-hardware state 0)]
+        (card-ability state :runner smc1 0)
+        (click-prompt state :runner (find-card "Self-modifying Code" (:deck (get-runner))))
+        (is (= 3 (:credit (get-runner))) "Paid 2 for SMC, 0 for install - 3 credits left")
+        (is (zero? (count(:hand (get-runner)))) "Runner hand should be empty")
+        (is (nil? (refresh qchip )) "Q chip should be trashed"))))
```
Comment (lostgeek) [inline test/clj/game_test/cards/hardware.clj]:
```suggestion
        (is (nil? (refresh qchip)) "Q chip should be trashed"))))
```

File: test/clj/game_test/cards/hardware.clj [hunk 3]
```diff
@@ -2217,6 +2217,55 @@
...
+        (play-from-hand state :runner "Q-Coherence Chip")
+        (is (= 5 (core/available-mu state)) "Gain 1 memory")))
+  (testing "Basic trash test"
+    (do-game
+      (new-game {:runner {:deck [(qty "Self-modifying Code" 3) "Q-Coherence Chip"]}})
+      (starting-hand state :runner ["Self-modifying Code" "Q-Coherence Chip"])
+      (take-credits state :corp)
+      (play-from-hand state :runner "Q-Coherence Chip")
+      (play-from-hand state :runner "Self-modifying Code")
+      (let [smc1 (get-program state 0),
```
Comment (lostgeek) [inline test/clj/game_test/cards/hardware.clj]:
```suggestion
      (let [smc1 (get-program state 0)
```

File: test/clj/game_test/cards/hardware.clj [hunk 4]
```diff
@@ -2217,6 +2217,55 @@
...
+            qchip (get-hardware state 0)]
+        (card-ability state :runner smc1 0)
+        (click-prompt state :runner (find-card "Self-modifying Code" (:deck (get-runner))))
```
Comment (lostgeek) [inline test/clj/game_test/cards/hardware.clj]:
```suggestion
        (click-prompt state :runner "Self-modifying Code")
```
The prompt is just showing text options. So you don't need to use `find-card`.

File: test/clj/game_test/cards/hardware.clj [hunk 5]
```diff
@@ -2217,6 +2217,55 @@
...
+        (is (= 3 (:credit (get-runner))) "Paid 2 for SMC, 0 for install - 3 credits left")
+        (is (zero? (count(:hand (get-runner)))) "Runner hand should be empty")
```
Comment (lostgeek) [inline test/clj/game_test/cards/hardware.clj]:
```suggestion
        (is (empty? (:hand (get-runner))) "Runner hand should be empty")
```

File: test/clj/game_test/cards/hardware.clj [hunk 6]
```diff
@@ -2217,6 +2217,55 @@
...
+    (do-game
+      (new-game {:corp {:deck [(qty "Breached Dome" 10)]}
+                 :runner {:deck ["Self-modifying Code", "Q-Coherence Chip"]}})               
+      (starting-hand state :runner ["Self-modifying Code" "Q-Coherence Chip"])
+      (starting-hand state :corp ["Breached Dome"])
+      (play-from-hand state :corp "Breached Dome" "New remote")
+      (take-credits state :corp)
+      (play-from-hand state :runner "Q-Coherence Chip")
+      (let [qchip (get-hardware state 0)]
+        (run-empty-server state "Server 1")
```
Comment (lostgeek) [inline test/clj/game_test/cards/hardware.clj]: Technically not an empty server, but I know this works. But we'll need to change this anyways once Noah's #4603 hits.

Comment (Kubik161) [reply test/clj/game_test/cards/hardware.clj]: copy & pasted this from Breached Dome tests. I have assumed that run-empty-server means "run server without ICE".

Comment (NoahTheDuke) [reply test/clj/game_test/cards/hardware.clj]: It's an empty server! Empty server means no ice in the tests. It'll advance the run to calling "successful-run". It's a good way of running a card to trash it or to trigger a replacement effect.

**Pull request merged**